src/traces/bar/plot.js

```diff
@@ -475,12 +476,51 @@ function getTransform(textX, textY, targetX, targetY, scale, rotate) {
     return transformTranslate + transformScale + transformRotate;
 }
 
-function getText(trace, index) {
-    var value = helpers.getValue(trace.text, index);
+function getText(calcTrace, index, fullLayout) {
+    var trace = calcTrace[0].trace;
+
+    var value;
+    if(!trace.textinfo || trace.textinfo === 'none') {
+        value = helpers.getValue(trace.text, index);
+    } else {
+        value = calcTextinfo(calcTrace, index, fullLayout);
+    }
+
     return helpers.coerceString(attributeText, value);
 }
 
 function getTextPosition(trace, index) {
     var value = helpers.getValue(trace.textposition, index);
     return helpers.coerceEnumerated(attributeTextPosition, value);
 }
+
+function calcTextinfo(calcTrace, index, fullLayout) {
+    var trace = calcTrace[0].trace;
+    var textinfo = trace.textinfo;
+    var cdi = calcTrace[index];
+
+    var separators = fullLayout.separators;
+    var parts = textinfo.split('+');
+    var hasFlag = function(flag) { return parts.indexOf(flag) !== -1; };
+    var thisText = [];
+
+    var delta = +cdi.rawS || cdi.s;
+    var final = cdi.v;
+    var initial = final - delta;
+
+    if(hasFlag('delta')) {
+        thisText.push(pieHelpers.formatPieValue(delta, separators));
+    }
+    if(hasFlag('final')) {
+        thisText.push(pieHelpers.formatPieValue(final, separators));
+    }
+    if(hasFlag('initial')) {
+        thisText.push(pieHelpers.formatPieValue(initial, separators));
+    }
+    if(hasFlag('text')) {
+        var tx = Lib.castOption(trace, cdi.i, 'text');
+        if(tx) thisText.push(tx);
+    }
+
+    return thisText.join('<br>'); // TODO: ' | ' may be a better dividor for bar
```

Comment:
Yep, it might be better to have `.join(' | ')` for horizontal waterfalls. But since I think most users will set only one `textinfo` flag, I doubt it's worth adding magic here.